client/src/providers/Search/Provider.tsx

```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import { useNavigate, useSearchParams } from 'react-router-dom';
+
+import { ItemsSearchOptions, itemsSearch } from '~/src/urls';
+
+import SearchContext from './context';
+import { Search, SearchFilters } from './types';
+
+interface SearchProviderProps {
+  children: React.ReactNode;
+}
+
+const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
+  const [filters, setFilters] = React.useState<SearchFilters>({});
+
+  const navigate = useNavigate();
+  const [urlSearchParams] = useSearchParams();
+
+  React.useEffect(() => {
+    const newFilters: SearchFilters = {};
+    const page = urlSearchParams.get('page');
+    if (page) {
+      newFilters.page = parseInt(page, 10);
+    }
+    const query = urlSearchParams.get('query');
+    if (query) {
+      newFilters.query = query;
+    }
+    const locationDistance = urlSearchParams.get('locationDistance');
+    const locationGooglePlaceID = urlSearchParams.get('locationGooglePlaceID');
+    if (locationGooglePlaceID) {
+      newFilters.location = { googlePlaceID: locationGooglePlaceID };
+      if (locationDistance) {
+        newFilters.location.distance = parseInt(locationDistance, 10);
+      }
+    }
+    const tagIDs = urlSearchParams.getAll('tag');
+    if (tagIDs) {
+      newFilters.tagIDs = tagIDs.map((tagID) => parseInt(tagID, 10));
+    }
+    setFilters(newFilters);
+  }, [setFilters, urlSearchParams]);
+
+  const onSearch = React.useCallback(
+    (newFilters: SearchFilters) => {
+      const itemsSearchOptions: ItemsSearchOptions = {};
+      if (newFilters.page) {
+        itemsSearchOptions.page = newFilters.page;
+      }
+      if (newFilters.query) {
+        itemsSearchOptions.query = newFilters.query;
+      }
+
+      const { location } = filters;
+      const { location: newLocation } = newFilters;
+      if (
+        (newLocation && newLocation.googlePlaceID) ||
+        (location && location.googlePlaceID)
+      ) {
+        itemsSearchOptions.location =
+          newLocation?.googlePlaceID || location!.googlePlaceID;
+        if (
+          (newLocation && newLocation.distance) ||
+          (location && location.distance)
+        ) {
```

Comment:
Does this also work?

```suggestion
        if (newLocation?.distance || location?.distance) {
```